base/src/Data/Macaw/AbsDomain/JumpBounds.hs

```diff
@@ -686,19 +764,29 @@ postJumpBounds :: forall arch ids
                 -> InitJumpBounds arch
 postJumpBounds cns regs = nextBlockBounds cns emptyBranchConstraints regs
 
--- | Get bounds for start of block after a branch (either the true or
--- false branch)
+data BranchBounds arch
+   = InfeasibleBranch
+   | TrueFeasibleBranch  !(InitJumpBounds arch)
+   | FalseFeasibleBranch !(InitJumpBounds arch)
+   | BothFeasibleBranch  !(InitJumpBounds arch) !(InitJumpBounds arch)
+
 postBranchBounds :: RegisterInfo (ArchReg arch)
                  => IntraJumpBounds arch ids -- ^ Bounds just before jump
-                 -> Value arch ids BoolType -- ^ Branch condition
-                 -> Bool -- ^ Flag indicating if branch predicate is true or false.
                  -> RegState (ArchReg arch) (Value arch ids)
                  -- ^ Register values at start of next state.
-                 -> Either String (InitJumpBounds arch)
-postBranchBounds cns c condIsTrue regs = do
-  brCns <- assertPred cns c condIsTrue
-  pure (nextBlockBounds cns brCns regs)
-
+                 -> Value arch ids BoolType -- ^ Branch condition
+                 -> BranchBounds arch
+postBranchBounds cns regs c = do
+  case (assertPred cns c True, assertPred cns c False) of
+    (Right brCnsT, Right brCnsF) ->
+      BothFeasibleBranch (nextBlockBounds cns brCnsT regs)
+                         (nextBlockBounds cns brCnsF regs)
+    (Right brCnsT, Left _) ->
+      TrueFeasibleBranch  (nextBlockBounds cns brCnsT regs)
+    (Left _, Right brCnsF) ->
+      FalseFeasibleBranch (nextBlockBounds cns brCnsF regs)
+    (Left _, Left _) ->
+      InfeasibleBranch
```

Comment:
Is it ever safe for us to assert that a branch is infeasible, given that we might have incomplete control flow?

Reply:
In the event of incomplete control flow (specifically jumps from elsewhere to the present block) everything is potentially unsound.
I'm not sure there's huge value in these special cases, branches may be always fine.  On the other hand, I think the soundness checks are unlikely to occur frequently.